Nimble Python SDK Workflows

nimblesorge opened the pull request:
All workflows - with the exception of host attach/detach.

nimblesorge committed:
Nimble Python SDK Workflows

Signed-off-by: Brian Sorge <brian.sorge@hpe.com>

ar-india commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+from nimbleclient.v1 import Client, NimOSAuthenticationError, NimOSAPIError
+
+
+hostname = 'brian-va.vlab.nimblestorage.com'
```
it may be a good idea to implement a config file and read the array details from that file. this will give flexibility to end user to just edit the config file when they run workflow against different array

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+import sys
```
Should we put these in separate files?  It might be easier for a user to navigate and find the workflow of interest.

ar-india replied:
agree

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+        ig = client.initiator_groups.update(id=ig.attrs['id'], description='add initiators',
+                                            iscsi_initiators=initiators)
+        screen('\tAdded initiator: {}'.format(ig.attrs['iscsi_initiators'][0]['label']), noisy)
+        # Delete an initiatior group
```
initiator

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+        screen('\tAdded initiator: {}'.format(ig.attrs['iscsi_initiators'][0]['label']), noisy)
+        # Delete an initiatior group
+        client.initiator_groups.delete(id=ig.id)
+        screen('\tDeleted initiatior group: {}'.format(ig.attrs['name']), noisy)
```
initiator

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+        # Create a volume
+        vol = client.volumes.create(name=vol_name, size=50, limit_iops=12000)
+        screen('\tCreated volume: {}, Id: {}'.format(vol.attrs['name'], vol.id), noisy)
+        # Create an initiatior group
```
initiator

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+        screen('\nWorkflow: Publish Volume - {}'.format(vol_id), noisy)
+        vol = client.volumes.get(id=vol_id)
+        vol_name = vol.attrs['name']
+        # Create an initiatior group
```
initiator

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+        screen('\nWorkflow: Protect Volume - {}'.format(vol_id), noisy)
+        # Get the volume
+        vol = client.volumes.get(vol_id)
+        # Create a volume volume_collection
```
volume volume_collection <= drop a volume :)

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+        screen('\tCreated protection schedule: {}, Id: {}'.format(ps.attrs['name'], ps.id), noisy)
+        client.protection_schedules.delete(id=ps.attrs['id'])
+        screen('\tDeleted protection schedule: {}'.format(ps.attrs['name']), noisy)
+        # Disssociate a volume with the volume collection and delete to volume collection
```
delete the volume collection

nimblesorge replied:
This happens on line 233, below.

rgcostea replied:
I was pointing out a grammatical issue:

> and delete to volume collection

should be "and delete **the** volume collection"

nimblesorge replied:
Gotcha :)

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+        screen('\tDeleted protection schedule: {}'.format(ps.attrs['name']), noisy)
+        # Disssociate a volume with the volume collection and delete to volume collection
+        client.volumes.dissociate(id=vol.attrs['id'])
+        screen('\tDisaocciated volume: {}, with volume vollection: {}'.
```
Disassociated

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+            sys.exit(1)
+        # Execute Workflows
+        vol_name = 'wfvol'
+        initiatior_group_name = 'wfig'
```
initiator

rgcostea commented on workflows/nimble_workflows.py:
```diff
@@ -0,0 +1,311 @@
+        initiatior_group_name = 'wfig'
+        initiator_group_lifecycle(client, initiatior_group_name, noisy)
+        volume_clone_lifecycle(client, vol_name, noisy)
+        acess_control_record_lifecycle(client, vol_name, initiatior_group_name, noisy)
```
access

rgcostea reviewed (approved):
looks good

nimblesorge committed:
Updates for PR

Removed nimble_workflows.py and replaced it with individual executable files for each workflow. Added workflow_common.py for some commonly used routines. Added workflow_all.py to provide a method for executing all of the workflows. Added workflow_config.json. Fixed various, egregious, spelling errors.

Signed-off-by: Brian Sorge <brian.sorge@hpe.com>

rgcostea commented on workflows/access_control_record_lifecycle.py:
```diff
@@ -0,0 +1,38 @@
+import os
```
Can you add copyright headers to these examples please?

nimblesorge committed:
More updates for PR

Added cleanup flag to all workflows.  Added copyright headers to all workflows. Expanded config file.

Signed-off-by: Brian Sorge <brian.sorge@hpe.com>

nimblesorge committed:
Updates to workflows

Made "skip cleanup" the default operating mode for all workflows. Added a --cleanup flag to force cleanup.
Modified all create methods so that an error does not occur if the object already exists - it is simple returned with a message to the user.
Added a workflow to list objects on the array.
Renamed workflow initiator_group_lifecycle.py to create_and_update_initiator_group.py
Cleaned up and included the nimble_sdk_helper.

Signed-off-by: Brian Sorge <brian.sorge@hpe.com>

nimblesorge committed:
Modifications to support change to nimble-sdk

Signed-off-by: Brian Sorge <brian.sorge@hpe.com>

ar-india commented on workflows/workflow_config.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "hostname":"brian-va.vlab.nimblestorage.com",
+  "username":"admin",
```
can we mask the array details like username , password and hostname??..
